flow-model-generator/modules/flow-model-generator-core/src/main/java/io/ballerina/flowmodelgenerator/core/converters/JsonToTypeMapper.java

```diff
@@ -0,0 +1,689 @@
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package io.ballerina.flowmodelgenerator.core.converters;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.google.gson.JsonPrimitive;
+import io.ballerina.flowmodelgenerator.core.TypesManager;
+import io.ballerina.flowmodelgenerator.core.converters.exception.JsonToRecordConverterException;
+import io.ballerina.flowmodelgenerator.core.model.Member;
+import io.ballerina.flowmodelgenerator.core.model.NodeKind;
+import io.ballerina.flowmodelgenerator.core.model.TypeData;
+import org.apache.commons.lang3.StringUtils;
+import org.ballerinalang.langserver.commons.workspace.WorkspaceManager;
+
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static io.ballerina.flowmodelgenerator.core.converters.utils.JsonToRecordMapperConverterUtils.escapeIdentifier;
+import static io.ballerina.flowmodelgenerator.core.converters.utils.JsonToRecordMapperConverterUtils.getAndUpdateFieldNames;
+import static io.ballerina.flowmodelgenerator.core.converters.utils.JsonToRecordMapperConverterUtils.getExistingTypeNames;
+import static io.ballerina.flowmodelgenerator.core.converters.utils.ListOperationUtils.difference;
+import static io.ballerina.flowmodelgenerator.core.converters.utils.ListOperationUtils.intersection;
+
+/**
+ * Converts a JSON string to a Ballerina type definition.
+ *
+ * @since 1.0.0
+ */
+public class JsonToTypeMapper {
+
+    private static final Gson GSON = new Gson();
+
+    private static final String ARRAY_RECORD_SUFFIX = "Item";
+    private static final String JSON_KEYWORD = "json";
+    private static final String NEW_TYPE_NAME = "NewType";
+    private static final String STRING_TYPE = "string";
+    private static final String BOOLEAN_TYPE = "boolean";
+    private static final String DECIMAL_TYPE = "decimal";
+    private static final String INT_TYPE = "int";
+    private static final String ANYDATA_TYPE = "anydata";
+
+    private final Set<String> existingFieldNamesSet;
+    private final Map<String, String> updatedFieldNames;
+    private final Map<String, TypeDesc> typeDefinitions;
+
+    // Cache for primitive type detection
+    private final Map<String, String> primitiveTypeCache = new ConcurrentHashMap<>();
+
+    private final String typePrefix;
+    private final boolean allowAdditionalFields;
+    private final boolean asInlineType;
+
+    public JsonToTypeMapper(boolean allowAdditionalFields, boolean asInlineType, String typePrefix,
```

Comment:
Builder pattern would be a better option here given the scalability for more options in the future. But we can do that in a separate PR.

Reply:
Yeah we can consider that option. But now, the idea of having separate smaller constructs to use lesser memory when manipulating. So with builder, aren't we adding a indirect overhead? Also, Using a builder is more appropriate when there is more parameters or fields. WDYT?